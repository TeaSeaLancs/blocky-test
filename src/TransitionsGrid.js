import React, { useCallback } from 'react';

import Grid from './Grid';
import Block from './Block';

function getBlockAnimationProps({ action, length }) {
  const className = `animate animate--${action}`;
  const style = {};

  if (action === 'drop') {
    // Fun fact: You can use CSS variables in animations! This allows us to have one drop animation
    // defined in CSS and then set the distance to drop on a per-element basis.
    style['--drop-length'] = `${length * 100}%`;
  }

  return { className, style };
}

function AnimatedBlock({ transition, onAnimationEnd, ...props }) {
  const {
    block: { x, y, colour },
  } = transition;

  /* The animated blocks have a little backing area which is grey so that the animations display
   * on top of the already changed grid, rather than messing with the display of the basic grid.
   * This means we could take out the TransitionsGrid component entirely and everything would still
   * function fine, just no animations :)
   *
   * The downside to this, though, is that every drop animation needs to take the same time, regardless
   * of distance, which isn't really that accurate on gravity, but it does look better for game purposes!
   */
  const style = {
    backgroundColor: 'grey',
    gridColumn: x + 1, // CSS Grid references are 1-based
    gridRow: y + 1,
  };

  // We set a specific data-testid on blocks which have an animation end event so we can find them in tests
  const testid = onAnimationEnd ? 'animated-block--with-listener' : undefined;

  return (
    <div style={style}>
      <Block
        x={x}
        y={y}
        colour={colour}
        data-testid={testid}
        {...getBlockAnimationProps(transition)}
        onAnimationEnd={onAnimationEnd}
        {...props}
      />
    </div>
  );
}

// This is totally extra to the requirements of the exercise, but the animations made it easier for me
// (and the user) to visualise that the changes to the grid after clicking on a block that were happening were correct.
export default function TransitionsGrid({
  grid = [],
  transitions,
  onTransitionEnd,
}) {
  const width = grid.length;
  const height = grid[0].length;

  const onAnimationEnd = useCallback(onTransitionEnd);

  // This is a bit cheeky, we're assuming that the last transition will be the longest, so we
  // listen out for that last animation finishing to fire the onTransitionEnd event.
  const blocks = transitions.map((transition, i) => {
    /* It seems like more work to have the key for the animated block be the X,Y of the grid when i would
     * suffice, but by doing this we ensure that if the user rapidly clicks, React doesn't re-use the same blocks
     * which could cause weird animation errors, which it would do if it used i.
     */
    const {
      block: { x, y },
    } = transition;
    return (
      <AnimatedBlock
        key={`${x},${y}`}
        transition={transition}
        onAnimationEnd={i === transitions.length - 1 ? onAnimationEnd : null}
      />
    );
  });

  return (
    <Grid width={width} height={height} className="floating">
      {blocks}
    </Grid>
  );
}
