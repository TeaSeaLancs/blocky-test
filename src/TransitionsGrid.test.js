import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { renderHook, act } from '@testing-library/react-hooks';

import TransitionsGrid from './TransitionsGrid';

import createSampleGrid from './__fixtures__/SampleGrid';
import createUpdateTransitions from './__fixtures__/UpdateTransitions';

describe('TransitionsGrid', () => {
  let grid, transitions;

  beforeEach(() => {
    grid = createSampleGrid();
    transitions = createUpdateTransitions();
  });

  it('renders an animated transition block for each transition passed', () => {
    const { asFragment } = render(
      <TransitionsGrid grid={grid} transitions={transitions} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls the onTransitionEnd prop when animations are finished', () => {
    const onTransitionEnd = jest.fn();
    const { getByTestId, debug } = render(
      <TransitionsGrid
        grid={grid}
        transitions={transitions}
        onTransitionEnd={onTransitionEnd}
      />
    );

    const animatedBlock = getByTestId('animated-block--with-listener');

    expect(onTransitionEnd).not.toHaveBeenCalled();
    fireEvent.animationEnd(animatedBlock);
    expect(onTransitionEnd).toHaveBeenCalled();
  });
});
