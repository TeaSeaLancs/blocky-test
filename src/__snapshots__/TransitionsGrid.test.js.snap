// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransitionsGrid renders an animated transition block for each transition passed 1`] = `
<DocumentFragment>
  <div
    class="grid floating"
  >
    <div
      style="background-color: grey; grid-column: 3; grid-row: 3;"
    >
      <div
        class="block animate animate--remove"
        data-x="2"
        data-y="2"
        style="background-color: blue;"
      />
    </div>
    <div
      style="background-color: grey; grid-column: 3; grid-row: 4;"
    >
      <div
        class="block animate animate--remove"
        data-x="2"
        data-y="3"
        style="background-color: blue;"
      />
    </div>
    <div
      style="background-color: grey; grid-column: 4; grid-row: 4;"
    >
      <div
        class="block animate animate--remove"
        data-x="3"
        data-y="3"
        style="background-color: blue;"
      />
    </div>
    <div
      style="background-color: grey; grid-column: 5; grid-row: 4;"
    >
      <div
        class="block animate animate--remove"
        data-x="4"
        data-y="3"
        style="background-color: blue;"
      />
    </div>
    <div
      style="background-color: grey; grid-column: 5; grid-row: 5;"
    >
      <div
        class="block animate animate--remove"
        data-x="4"
        data-y="4"
        style="background-color: blue;"
      />
    </div>
    <div
      style="background-color: grey; grid-column: 5; grid-row: 3;"
    >
      <div
        class="block animate animate--remove"
        data-x="4"
        data-y="2"
        style="background-color: blue;"
      />
    </div>
    <div
      style="background-color: grey; grid-column: 5; grid-row: 2;"
    >
      <div
        class="block animate animate--drop"
        data-x="4"
        data-y="1"
        style="background-color: red;"
      />
    </div>
    <div
      style="background-color: grey; grid-column: 5; grid-row: 1;"
    >
      <div
        class="block animate animate--drop"
        data-x="4"
        data-y="0"
        style="background-color: green;"
      />
    </div>
    <div
      style="background-color: grey; grid-column: 4; grid-row: 3;"
    >
      <div
        class="block animate animate--drop"
        data-x="3"
        data-y="2"
        style="background-color: green;"
      />
    </div>
    <div
      style="background-color: grey; grid-column: 4; grid-row: 2;"
    >
      <div
        class="block animate animate--drop"
        data-x="3"
        data-y="1"
        style="background-color: yellow;"
      />
    </div>
    <div
      style="background-color: grey; grid-column: 4; grid-row: 1;"
    >
      <div
        class="block animate animate--drop"
        data-x="3"
        data-y="0"
        style="background-color: red;"
      />
    </div>
    <div
      style="background-color: grey; grid-column: 3; grid-row: 2;"
    >
      <div
        class="block animate animate--drop"
        data-x="2"
        data-y="1"
        style="background-color: green;"
      />
    </div>
    <div
      style="background-color: grey; grid-column: 3; grid-row: 1;"
    >
      <div
        class="block animate animate--drop"
        data-x="2"
        data-y="0"
        style="background-color: yellow;"
      />
    </div>
  </div>
</DocumentFragment>
`;
